core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -335,20 +335,34 @@ private[chisel3] class DynamicContext(
   val throwOnFirstError: Boolean) {
   val importDefinitionAnnos = annotationSeq.collect { case a: ImportDefinitionAnnotation[_] => a }
 
-  // Ensure there are no repeated names for imported Definitions
-  val importDefinitionNames = importDefinitionAnnos.map { a => a.definition.proto.name }
-  if (importDefinitionNames.distinct.length < importDefinitionNames.length) {
-    val duplicates = importDefinitionNames.diff(importDefinitionNames.distinct).mkString(", ")
+  // Map holding the actual names of extModules
+  // Pick the definition name by default in case not passed through annotation.
+  val importDefinitionMap = importDefinitionAnnos
+    .map(a => ((a.definition.proto.name, a.overrideDefName.getOrElse(a.definition.proto.name))))
+    .toMap
+
+  // Ensure there are no repeated names for imported Definitions - both Proto Names as well as ExtMod Names
+  val importAllDefinitionProtoNames = importDefinitionAnnos.map { a => a.definition.proto.name }
+  val importDistinctDefinitionProtoNames = importDefinitionMap.keys.toSeq
+  val importAllDefinitionExtModNames = importDefinitionMap.toSeq.map(_._2)
+  val importDistinctDefinitionExtModNames = importAllDefinitionExtModNames.distinct
+
+  if (importDistinctDefinitionProtoNames.length < importAllDefinitionProtoNames.length) {
+    val duplicates = importAllDefinitionProtoNames.diff(importDistinctDefinitionProtoNames).mkString(", ")
     throwException(s"Expected distinct imported Definition names but found duplicates for: $duplicates")
   }
+  if (importDistinctDefinitionExtModNames.length < importAllDefinitionExtModNames.length) {
+    val duplicates = importAllDefinitionExtModNames.diff(importDistinctDefinitionExtModNames).mkString(", ")
+    throwException(s"Expected distinct overrideDef names but found duplicates for: $duplicates")
+  }
```

Comment:
Can you express this in such a way that it doesn't add 5 public fields to this class (which means each of these must stick around for all of Chisel elaboration). `importDefinitionMap` seems to be the only one that is needed outside of the class, can the rest be local vals of a `def` or block?

Reply:
@jackkoenig  - good point! Let me refactor.